src/main/java/net/imglib2/img/display/imagej/AbstractVirtualStack.java

```diff
@@ -340,27 +358,86 @@ public String toString()
+	public float[] getVoxels( final int x0, final int y0, final int z0, final int w, final int h, final int d, float[] voxels, final int channel )
 	{
-		throw new UnsupportedOperationException();
+		return accessVoxels( x0, y0, z0, w, h, d, voxels, channel, false );
 	}
 
 	@Override
 	public void setVoxels( final int x0, final int y0, final int z0, final int w, final int h, final int d, final float[] voxels )
 	{
-		throw new UnsupportedOperationException();
+		accessVoxels( x0, y0, z0, w, h, d, voxels, null, true );
 	}
 
 	@Override
 	public void setVoxels( final int x0, final int y0, final int z0, final int w, final int h, final int d, final float[] voxels, final int channel )
 	{
-		throw new UnsupportedOperationException();
+		accessVoxels( x0, y0, z0, w, h, d, voxels, channel, true );
+	}
+
+	private float[] accessVoxels( int x0, int y0, int z0, int w, int h, int d, float[] voxels, Integer optionalChannel, boolean setVoxel )
+	{
+		checkBounds( x0, y0, z0, w, h, d );
+		if( ! setVoxel )
+			voxels = checkResultArray( w, h, d, voxels );
+		BiConsumer< Object, FloatType > action = ( BiConsumer< Object, FloatType > ) voxelAccessAction( optionalChannel, setVoxel );
+		loopOverVoxels( x0, y0, z0, w, h, d, voxels, action );
+		return voxels;
+	}
+
+	private void checkBounds( int x0, int y0, int z0, int w, int h, int d )
+	{
+		boolean inBounds = (x0 >= 0) && (x0 + w <= width) && (y0 >= 0) && (y0 + h <= height) && (z0 >= 0) && (z0 + d <= size);
+		if(! inBounds)
+			throw new IndexOutOfBoundsException();
+	}
+
+	private float[] checkResultArray( int w, int h, int d, float[] voxels )
+	{
+		if ( voxels == null || voxels.length != w * h * d )
+			voxels = new float[ w * h * d ];
+		return voxels;
+	}
+
+	private void loopOverVoxels( int x0, int y0, int z0, int w, int h, int d, float[] voxels, BiConsumer< Object, FloatType > action )
+	{
+		FinalInterval interval = Intervals.createMinSize( x0, y0, w, h );
+		Cursor< FloatType > output = ArrayImgs.floats( voxels, w, h, d ).cursor();
+		for ( int z = z0 + offset; z < z0 + offset + d; z++ )
+		{
+			Cursor< ? > cursor = Views.flatIterable( Views.interval( getSliceZeroBasedIndex( z ), interval ) ).cursor();
+			while ( cursor.hasNext() ) {
+				action.accept( cursor.next(), output.next() );
+			}
+		}
+	}
+
+	private BiConsumer<?, FloatType> voxelAccessAction( Integer channel, boolean isSetVoxels )
+	{
+		if ( bitDepth == 24 )
+			return arbgVoxelAccessAction( channel, isSetVoxels );
+		return isSetVoxels ?
+				(BiConsumer< RealType, FloatType >) (o, i) -> o.setReal( i.getRealFloat() ) :
+				(BiConsumer< RealType, FloatType >) (i, o) -> o.setReal( i.getRealFloat() );
+	}
+
+	private static BiConsumer<ARGBType, FloatType> arbgVoxelAccessAction( Integer channel, boolean isSetVoxel )
+	{
+		if( channel == null )
+			return isSetVoxel ?
+					( o, i ) -> o.set( (int) i.getRealFloat() ) :
+					( i, o ) -> o.set( i.get() );
```

Comment:
I would find this more readable, if variables are renamed like
```java
return isSetVoxel ?
		( a, b ) -> a.set( ( int ) b.getRealFloat() ) :
		( a, b ) -> b.set( a.get() );
```
I can mentally assign "Ok, `a` that's the ARGBType, and `b` that's the `FloatType`" and then I parse the lines.
Using suggestive names ('i' "input", `o` "output") and then flipping the order is much more confusing (at least to me).